### ISSUE TITLE
Bind CoapServer to listen to multicast address does not work

### ISSUE THREAD

**Issue by lightertu** [eclipse/californium#219]
I found out that in the Request class there is an option to set multicast, but how can I ask the server to listen to the multicast address? Also If I am using the CoapClient class to send requests how can I add multicast options?

I have tried the following code
```java
CoapServer server = ...
InetAddress addr = InetAddress.getByName("224.0.1.187");
bindToAddress = new InetSocketAddress(addr, COAP_PORT);
multicast = new CoapEndpoint(bindToAddress);
server.addEndpoint(multicast);
```

then got this error
```
SEVERE: Cannot start server endpoint [/224.0.1.187:5683]
java.net.BindException: Cannot assign requested address: Cannot bind
	at java.net.DualStackPlainDatagramSocketImpl.socketBind(Native Method)
	at java.net.DualStackPlainDatagramSocketImpl.bind0(DualStackPlainDatagramSocketImpl.java:84)
	at java.net.AbstractPlainDatagramSocketImpl.bind(AbstractPlainDatagramSocketImpl.java:93)
	at java.net.DatagramSocket.bind(DatagramSocket.java:392)
	at java.net.DatagramSocket.<init>(DatagramSocket.java:242)
	at java.net.DatagramSocket.<init>(DatagramSocket.java:299)
	at org.eclipse.californium.elements.UDPConnector.start(UDPConnector.java:113)
	at org.eclipse.californium.core.network.CoapEndpoint.start(CoapEndpoint.java:337)
	at org.eclipse.californium.core.CoapServer.start(CoapServer.java:200)
	at org.netsec.iottestbed.nonembedded.Server.start(Server.java:88)
	at org.netsec.iottestbed.nonembedded.Server.main(Server.java:98)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```

UPDATE:

Previous errors occurred only on windows, I also tried the same thing mac, it shows that 
```
Feb 21, 2017 5:48:32 PM org.eclipse.californium.core.network.CoapEndpoint start
INFO: Starting endpoint at /ff05:0:0:0:0:0:0:fd:5683
```

However, if I send a request to the multicast address using a client on the same machine it won't respond. I also tried to send requests to other interfaces they all worked.

**Comment by vikram919** [comment]
@lightertu 
Hi did you figure it out how to solve the problem?
I have a solution, Binding a multicast address to coap server can be on UDPConnector level you can use the coap default port number and replace a DatagramSocket to MulticastSocket. 
After creating the multicast socket you can select which network interface to listen for multicast requests, then use `socket.joinGroup(InetAddress.getByName("224.0.1.187"));` to receive the client request multicast requests.

**Comment by sophokles73** [comment]
Multicast is not a supported feature in Californium (yet). There are some places in the code that suggest otherwise, but in fact, there is currently nobody working on implementing RFC compliance.
As @vikram919  pointed out, this would probably require the UDPConnector to bind to a `MulticastSocket` instead of a plain `DatagramSocket` based on the outcome of `InetAddress.isMulticastAddress()`. Maybe somebody wants to create a PR for that?

**Comment by vikram919** [comment]
@sophokles73 
Is this what you mean?

```
                 if (localAddr.getAddress().isMulticastAddress()) {
			socket = new MulticastSocket(localAddr.getPort());
                        ((MulticastSocket) socket).joinGroup(localAddr.getAddress());
                  } else {
			socket = new DatagramSocket(localAddr.getPort(), localAddr.getAddress());
		  }
```

**Comment by sophokles73** [comment]
I think so. However, I am by no means an expert on multicast IP. So you should do some tests, if this actually does the trick ...

**Comment by vikram919** [comment]
There are few issues I did not understand that the specification RFC 7390 (Group Communications for the CoAP) or RFC 7252  did not explain:

**1. Token matching after receiving the response incontext to multicast request:** californium client does not know which group or endpoint has sent the response for the multicast request sent earlier from client endpoint, the californium 2.0 branch implements token matching with respect to response received at 

`californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java` 
at method: `fromInboundMessage(final Message message)`
`message.getSourceContext().getPeerAddress(); `

and cross checks the token obtained from this peer address to the token generated while sending a     request to the same peer adress before. In case of multicast request the Destination peer address is a multicast address so the token created using  multicast adress doesn't match to response token generated from response source address.

For now I have introduced a condtion if 
```
if(response.getPayload().contains("/.well-known/core")){
PeerAddress = InetAddress.getHostByName(COAP_DEFAULT_MC_ADDRESS);
}
```
or
This check has to be relaxed incase of Multicast requests and responses.
If anyone know how this can also be solved, comments are welcome

**Comment by vikram919** [comment]
According to RFC 7252, Section 8.2. 

>  When matching a response to a multicast request, only the token MUST
   match; the source endpoint of the response does not need to (and will
   not) be the same as the destination endpoint of the original request.

Based on this information I have implemented the condition to check only token for the incoming unicast responses for the multicast request.
Now am writing some unit tests and at the end hopefully I can make a pull request.

**Comment by vikram919** [comment]
@lightertu 
https://github.com/vikram919/californium/blob/udp_multicast_2_0_x/demo-apps/cf-helloworld-server/src/main/java/org/eclipse/californium/examples/MulticastTestServer.java

Please check out this branch for testing multicast.

**Comment by boaks** [comment]
If (still) interested, check PR #742.

**Comment by boaks** [comment]
Fixed with merge of #742
